pkg/ruler/ruler.go

```diff
@@ -507,8 +536,24 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
 			}
 		}
 	} else {
-		// Even if the replication factor is set to a number bigger than 1, only the first ruler evaluates the rule group
 		ownsRuleGroup = rlrs.Instances[0].Addr == instanceAddr
+		if r.Config().EnableHA {
+			for i, ruler := range rlrs.Instances {
+				if ruler.Addr == instanceAddr && i == 0 {
+					level.Debug(r.Logger()).Log("msg", "primary taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "token", hash, "ruler", instanceAddr)
+					ownsRuleGroup = true
+					break
+				}
+				if ruler.Addr == instanceAddr && nonPrimaryInstanceOwnsRuleGroup(r, g, rlrs.GetAddresses()[:i], instanceAddr) {
+					level.Info(r.Logger()).Log("msg", "non-primary ruler taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "token", hash, "ruler", instanceAddr)
+					ownsRuleGroup = true
+					break
+				}
+			}
+		} else {
+			// Even if the replication factor is set to a number bigger than 1, only the first ruler evaluates the rule group
+			ownsRuleGroup = rlrs.Instances[0].Addr == instanceAddr
```

Comment:
Is this else statement necessary? I see we still keep L539 above

Reply:
Good catch! The `else` is necessary but L539 must be removed

Reply:
does this needs to be inside the for??? can we on foreach the instances if `EnableHAEvaluation` is enabled and if not, we do `rlrs.Instances[0].Addr == instanceAddr` check?

Reply:
@alanprot It already is that way, no? if `EnableHAEvaluation` is disabled, we do `rlrs.Instances[0].Addr == instanceAddr` check outside the for loop. If `EnableHAEvaluation` is enabled, then loop thru the instances and perform the liveness check